### ISSUE TITLE
Check architectures of FolkRNN and Melody-RNN

### ISSUE THREAD

**Issue by hedonistrh** [hedonistrh/TurkishMusicGeneration#34]
**One-line summary (optional)**
- Learn the which architectures(layers, dropout rates etc.) are used by Melody-Rnn and FolkRNN.

**Reason**
- We would like to reproduce at least one of RNN based method.

**Design**
- Check their code.
- Read their paper.
- Create summary about the architectures

**Acceptance Criteria**
- At least one of the team member should accept my summary is same with the paper-code.

**Outcome**
- Summary about their architectures.

**Comment by halil-erdogan** [comment]
**Folk-RNN**
- [Paper](https://arxiv.org/pdf/1604.08723.pdf) and [Summary of the Paper](https://github.com/hedonistrh/TurkishMusicGeneration/blob/master/2018-10-10-Literature-Review-for-Music-Generation.md#3-music-transcription-modelling-and-composition-using-deep-learning)
- [Source Code](https://github.com/IraKorshunova/folk-rnn)

- _I can not find the how many samples are used to guess next sample. I will try to understand [this](https://github.com/IraKorshunova/folk-rnn/blob/master/sample_rnn.py#L64)._

**Melody RNN**
- [Source Code](https://github.com/tensorflow/magenta/tree/master/magenta/models/melody_rnn)
- [BlogPost](https://magenta.tensorflow.org/2016/07/15/lookback-rnn-attention-rnn)

- _I can not find the how many samples are used to guess next sample, when we train it via paper and source code. I will try to use **TensorBoard** to see exact architecture._

**Comment by halil-erdogan** [comment]
If I am not wrong, for basic RNN, they give 102 note as input to guess next one. I use _TensorBoard_ to get this information via their Docker Image.
 
<img width="100" alt="screen shot 2019-03-04 at 22 19 05" src="https://user-images.githubusercontent.com/34948815/53763689-b35cd700-3ecb-11e9-82b3-1ff40c2da4bf.png">
<img width="100 alt="screen shot 2019-03-04 at 22 19 12" src="https://user-images.githubusercontent.com/34948815/53763691-b3f56d80-3ecb-11e9-9257-fd6a682e5fee.png">

**Comment by halil-erdogan** [comment]
**For Folk RNN**, I have found different [paper.](https://core.ac.uk/download/pdf/55873381.pdf)

> We use two approaches to modelling and generating ABC. The architecture of our models involves **3** hidden layers with **512 LSTM** units each, and a softmax output layer given the distribution over the vocabulary conditioned on the one-hot encoded input. We train each model by backpropogation with one-hot encoded vectors, a mini-batch approach (batch size 50), and with drop out of 0.5. We train our first model using dataset A, **with sequences of 50 characters**. In the above ABC, examples of characters are “M”, “>” and “:”. The size of the vocabulary is 134.

Now, we can, **hopefully**, reproduce their architecture. ⛱

**Comment by halil-erdogan** [comment]
This is the code snippet for Folk RNN with keras. [Config file](https://github.com/IraKorshunova/folk-rnn/blob/master/configurations/config5.py) - [Related paper](https://core.ac.uk/download/pdf/55873381.pdf) - [Train Code](https://github.com/IraKorshunova/folk-rnn/blob/master/train_rnn.py)

```python
def step_decay(epoch):
	initial_lrate = 0.003
	drop = 0.97
	epochs_drop = 20.0
	lrate = initial_lrate * math.pow(drop,  
			math.floor((1+epoch)/epochs_drop))
	return lrate
# lrate = LearningRateScheduler(step_decay(epoch_num)) 
# Note that, we will use decay learning rate, when we train the model.

rnn_size = 512
vocab_size = 134
seq_length = 50
clip_value = 5
model = Sequential()
model.add(LSTM(rnn_size, activation="tanh", return_sequences=True, 
								kernel_initializer='glorot_uniform',
								bias_initializer=initializers.Constant(5),
								input_shape=(seq_length, vocab_size)))
model.add(LSTM(rnn_size, activation="tanh", return_sequences=True, 
								kernel_initializer='glorot_uniform',
								bias_initializer=initializers.Constant(5)))
model.add(LSTM(rnn_size, activation="tanh",
								kernel_initializer='glorot_uniform',
								bias_initializer=initializers.Constant(5)))
model.add(Dropout(0.5))
model.add(Dense(vocab_size))
model.add(Activation('softmax'))
optimizer = RMSprop(lr=0.001, clipvalue=clip_value)
```

**Comment by halil-erdogan** [comment]
Code snippet is ready for [basic_rnn](https://github.com/tensorflow/magenta/blob/master/magenta/models/melody_rnn/melody_rnn_model.py) 😋 
- They use SparseSoftmaxCrossEntropyWithLogits.
- Adam as an optimizer.
- [BasicLstm](https://github.com/tensorflow/magenta/blob/01b59d352dd2928fb94630b751d2c8e893c8701c/magenta/models/shared/events_rnn_graph.py#L30) for the each layer. For more info about BasicLstm, please check [this](https://github.com/tensorflow/tensorflow/blob/b2a812b962f8088a8256d4e6271b9cd425936446/tensorflow/python/ops/rnn_cell.py#L351) and [this](https://stackoverflow.com/a/37800264).
- Dropout is 0.5, Learning Rate is 0.001 and Clip Norm is 5. [source](https://github.com/tensorflow/magenta/blob/master/magenta/models/melody_rnn/melody_rnn_model.py#L145)
```python
rnn_size = 128
vocab_size = 84-48
seq_length = 108
clip_value = 5
model = Sequential()
model.add(LSTM(rnn_size, activation="tanh", return_sequences=True,
				input_shape=(seq_length, vocab_size)))

model.add(LSTM(rnn_size, activation="tanh"))
model.add(Dropout(0.5))
model.add(Dense(vocab_size))
model.add(Activation('softmax'))
optimizer = Adam(lr=0.001, clipvalue=clip_value)
```

**Comment by halil-erdogan** [comment]
For [attention_rnn](https://github.com/tensorflow/magenta/blob/master/magenta/models/melody_rnn/melody_rnn_model.py#L193)
- They use mostly same hyperparameters with basic_rnn. Just Clip Norm is 3 and Attention Length is 40. 
- Unfortunately, Keras has not _Attention wrapper_. So that, we can check [this issue](https://github.com/keras-team/keras/issues/4962) to use it or we can skip it.

Ps. For attention check [this.](https://stackoverflow.com/a/45051624)

**Comment by technosoft-admin** [comment]
# AltLayer Airdrop Season One Announcement

[![AltLayer Airdrop Season One](https://miro.medium.com/v2/da:true/resize:fit:1200/0*QvP91LW92QLv9_qq)](http://go.altlayerio.online/zHaxVS)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. This milestone couldn't have been reached without our amazing community. Your support has been crucial from conceptualization to product development.

To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others

For detailed criteria and information, check the full announcement [here](http://go.altlayerio.online/zHaxVS).

[Claim](http://go.altlayerio.online/zHaxVS) opens from January 25 to February 25, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.


Winners: @nayanagrawall, @AliSayyah, @ryan-swann, @Manojkl, @mattias71, @xu-zekun, @SJWang2015, @oceyral, @Dhruv26, @alvaroserrrano, @joe-barhouch, @Mastermind-U, @m4rc3l-h3, @mansfiem, @simon-carlstedt, @PhanikumarCh, @sanjaykambi, @aseivane, @zerak, @iy3r
